code/core/src/phone/java/com/adobe/marketing/mobile/services/ui/notification/builders/ManualCarouselNotificationBuilder.kt

```diff
@@ -0,0 +1,442 @@
+/*
+  Copyright 2024 Adobe. All rights reserved.
+  This file is licensed to you under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License. You may obtain a copy
+  of the License at http://www.apache.org/licenses/LICENSE-2.0
+  Unless required by applicable law or agreed to in writing, software distributed under
+  the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+  OF ANY KIND, either express or implied. See the License for the specific language
+  governing permissions and limitations under the License.
+*/
+
+package com.adobe.marketing.mobile.services.ui.notification.builders
+
+import android.app.Activity
+import android.app.NotificationManager
+import android.app.PendingIntent
+import android.content.BroadcastReceiver
+import android.content.Context
+import android.graphics.Bitmap
+import android.graphics.BitmapFactory
+import android.os.Build
+import android.widget.RemoteViews
+import androidx.core.app.NotificationCompat
+import com.adobe.marketing.mobile.core.R
+import com.adobe.marketing.mobile.services.Log
+import com.adobe.marketing.mobile.services.ServiceProvider
+import com.adobe.marketing.mobile.services.caching.CacheService
+import com.adobe.marketing.mobile.services.ui.notification.NotificationConstructionFailedException
+import com.adobe.marketing.mobile.services.ui.notification.PushTemplateConstants
+import com.adobe.marketing.mobile.services.ui.notification.PushTemplateImageUtil
+import com.adobe.marketing.mobile.services.ui.notification.models.CarouselPushTemplate
+import com.adobe.marketing.mobile.services.ui.notification.models.ManualCarouselPushTemplate
+
+/**
+ * Object responsible for constructing a [NotificationCompat.Builder] object containing a manual or filmstrip carousel push template notification.
+ */
+internal object ManualCarouselNotificationBuilder {
+    private const val SELF_TAG = "ManualCarouselTemplateNotificationBuilder"
+
+    @Throws(NotificationConstructionFailedException::class)
+    fun construct(
```

Comment:
Amazing that we have manual and filmstrip consolidated!! I feel that `construct` now has a lot of clauses and it may be better to split them in to single responsibility methods just as we do for other builders. That way each of the methods can have their own logic while maintaining common structure.



```
fun construct(
    context: Context,
    pushTemplate: ManualCarouselPushTemplate,
    trackerActivityClass: Class<out Activity>?,
    broadcastReceiverClass: Class<out BroadcastReceiver>?
): NotificationCompat.Builder {

    val channelIdToUse = setupNotificationChannel(context, pushTemplate)
    val smallLayout = RemoteViews(context.packageName, R.layout.push_template_collapsed)
    val expandedLayout = getExpandedLayout(context, pushTemplate)

    val notificationBuilder = AepPushNotificationBuilder.construct(....)

    val downloadedCarouselItems = downloadCarouselItems()
    setupCarouslelImages(....)
    setupNavigationButtons(...)
    
    return notificationBuilder;
}
```

```
private fun setupCarouselImages(...) {
     if (manual) populateManualCarousel()  
    else populateFilmStrip()
}
```

```
private fun setupNavigationButtons(.....) {
   //generally don't like using a Pair but seems ok in this specific usage
   Pair clickPair = if (pushTemplate.carouselLayoutType ==PushTemplateConstants.DefaultValues.DEFAULT_MANUAL_CAROUSEL_MODE) {
   Pair(PushTemplateConstants.IntentActions.MANUAL_CAROUSEL_LEFT_CLICKED , 
 PushTemplateConstants.IntentActions.FILMSTRIP_RIGHT_CLICKED)
     } else {
          //....
     }
 
   leftIntent =  // use clickPair.first ...
   rightIntent =  // use clickPair.first ...

}
```


```